packages/rpc-transport-http/src/json-parse-with-bigint.ts

```diff
@@ -0,0 +1,76 @@
+/**
+ * This function is a replacement for `JSON.parse` that can handle large integers by
+ * parsing them as BigInts. It will only transform integers that are too large to be
+ * represented as a regular JavaScript number.
+ */
+export function jsonParseWithLargeIntegersAsBigInts(json: string): unknown {
+    return JSON.parse(wrapLargeIntegers(json), (_, value) => {
+        return isWrappedBigInt(value) ? unwrapBigInt(value) : value;
+    });
+}
+
+/**
+ * This function takes a JSON string and wraps any large, unsafe integers in a special
+ * format that can be later recognized and unwrapped by a `JSON.parse` reviver.
+ */
+export function wrapLargeIntegers(json: string): string {
+    return transformUnquotedSegments(json, unquotedSegment => {
+        return unquotedSegment.replaceAll(/(-?\d+)/g, (_, value: string) => {
+            return Number.isSafeInteger(Number(value)) ? value : wrapBigInt(value);
+        });
+    });
+}
+
+/**
+ * This function takes a JSON string and transforms any unquoted segments using the provided
+ * `transform` function. This means we can be sure that our transformations will never occur
+ * inside a double quoted string — even if it contains escaped double quotes.
+ */
+function transformUnquotedSegments(json: string, transform: (value: string) => string): string {
+    /**
+     * This regex matches any part of a JSON string that isn't wrapped in double quotes.
+     *
+     * For instance, in the string `{"age":42,"name":"Alice \"The\" 2nd"}`, it would the
+     * following parts: `{`, `:42,`, `:`, `}`. Notice the whole "Alice \"The\" 2nd" string
+     * is not matched as it is wrapped in double quotes and contains escaped double quotes.
+     *
+     * The regex is composed of two parts:
+     *
+     *   1. The first part `^([^"]+)` matches any character until we reach the first double quote.
+     *   2. The second part `("(?:\\"|[^"])+")([^"]+)` matches any double quoted string that may
+     *      and any unquoted segment that follows it. To match a double quoted string, we use the
+     *      `(?:\\"|[^"])` regex to match any character that isn't a double quote whilst allowing
+     *      escaped double quotes.
+     */
+    const unquotedSegmentsRegex = /^([^"]+)|("(?:\\"|[^"])+")([^"]+)/g;
+
+    return json.replaceAll(unquotedSegmentsRegex, (_, firstGroup, secondGroup, thirdGroup) => {
+        // If the first group is matched, it means we are at the
+        // beginning of the JSON string and we have an unquoted segment.
+        if (firstGroup) return transform(firstGroup);
+
+        // Otherwise, we have a double quoted string followed by an unquoted segment.
+        return `${secondGroup}${transform(thirdGroup)}`;
+    });
+}
```

Comment:
I would be really tempted to implement this as a parser instead of a regex. Walking every character:

1. Toggle `inQuote` on and off when you encounter a non-escaped `"`.
2. If `inQuote` is `false` and you encounter a contiguous run of numeric digits without a `.` in the middle, then wrap 'em.

```ts
const str = '{"hi":9223372036854775807,"there":123.4,"bye":"content is \\"1337\\""}';
let out = [];
let inQuote = false;
for (let ii = 0; ii < str.length; ii++) {
    let isEscaped = false;
    if (str[ii] === '\\') {
        out.push(str[ii++]);
        isEscaped = !isEscaped;

    }
    if (ii >= str.length) {
        break;
    }
    if (str[ii] === '"') {
        out.push(str[ii++]);
        if (!isEscaped) {
            inQuote = !inQuote;
        }
    }
    if (ii >= str.length) {
        break;
    }
    if (!inQuote) {
        let consumedNumber = '';
        while (str[ii].match(/[\d\.]/)) {
            consumedNumber += str[ii++];
        }
        if (consumedNumber.length) {
            if (consumedNumber.includes('.')) {
                out.push(consumedNumber);
            } else {
                out.push(`{"$n":"${consumedNumber}"}`)
            }
        }
    }
    out.push(str[ii]);
}
console.log(out.join(''));
```

Reply:
I don't disagree. Regexes always feel a bit hacky but at the same time I was struggling to come up with examples that would not work with this implementation so I though it might be the least intrusive way to implement this. That being said, we still need a shit loads more test to make sure we're not messing up with the integrity of the response data.

I'm up for giving the parser a go though. I don't know if you've seen [this library](https://github.com/sidorares/json-bigint/blob/master/lib/parse.js) but it could be useful to get some inspiration from. Tbh, I started looking at this parsing logic first and though: hmm maybe a parser is more complicated than it looks like and moved to regexes instead haha. But I think this library does a bit too much for our need anyway.

Anyway, to summary my chaotic thoughts, we can either:
1. Continue to explore regexes.
2. Move towards a parsing solution that only transforms the string (what you're suggesting).
3. Move towards a full parsing solution that outputs parsed data directly by forking/editing [this library](https://github.com/sidorares/json-bigint/blob/master/lib/parse.js).

In any case, we need lots more tests.

I lean towards solution 2, it seems like a clean and robust approach.

Reply:
String-to-string parsing will be way more resilient, I think. It's already resilient to annoying shit like newlines, spaces, and infinite `\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\` escape sequences, without added complexity.

Reply:
I've implemented solution 2 in two different ways:
- One that uses a simple regex to consume numbers.
- One that uses plain loops to consume numbers.

The `while (str[ii]?.match(/[\d\.Ee]/))` wasn't enough to correctly match numbers — e.g. it was matching `true` because of the `e` and returning `{"result":tru{"$n":"e"}}`.

I've also left the old regex-only implementation to compare in the benchmarks but I'm happy to remove it because the parsing is much more robust.